tests/e2e/common/diagnosticcollector.go

```diff
@@ -0,0 +1,564 @@
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	nfdclient "sigs.k8s.io/node-feature-discovery/pkg/generated/clientset/versioned"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	// the core group
+	Pods       = "pods"
+	Nodes      = "nodes"
+	Namespaces = "namespaces"
+
+	// the apps group
+	Deployments = "deployments"
+	DaemonSets  = "daemonsets"
+
+	// the batch group
+	Jobs = "jobs"
+
+	// Supported extensions
+	NodeFeature     = "nodeFeature"
+	NodeFeatureRule = "nodeFeatureRule"
+)
+
+type DiagnosticCollector struct {
+	Objects
+
+	NfdClient *nfdclient.Clientset
+
+	clientset   *kubernetes.Clientset
+	artifactDir string
+	namespace   string
+	collectors  []collectFunc
+	logFile     *os.File
+	errLogFile  *os.File
+}
+
+// A type to wrap the DiagnosticCollector items core, apps, batch, and extensions
+type Objects struct {
+	Core       []string
+	Apps       []string
+	Batch      []string
+	Extensions []string
+}
+
+type collectFunc func(ctx context.Context) error
+
+func NewDiagnosticCollector(namespace, artifactDir string) (*DiagnosticCollector, error) {
+	dc := &DiagnosticCollector{}
+
+	// Set the namespace
+	dc.namespace = namespace
+
+	// Create the ARTIFACT_DIR
+	dc.artifactDir = artifactDir
+	if err := os.MkdirAll(artifactDir, os.ModePerm); err != nil {
+		return dc, err
+	}
+
+	// Redirect stdout and stderr to logs
+	logFile, err := os.Create(filepath.Join(artifactDir, "diagnostic_collector.log"))
+	if err != nil {
+		return dc, err
+	}
+	dc.logFile = logFile
+
+	errLogFile, err := os.Create(filepath.Join(artifactDir, "diagnostic_collector.stderr.log"))
+	if err != nil {
+		return dc, err
+	}
+	dc.errLogFile = errLogFile
+
+	// Get the kubeconfig
+	kubeconfig := os.Getenv("KUBECONFIG")
+
+	// Create the Kubernetes client
+	clientset, err := createKubernetesClient(kubeconfig)
+	if err != nil {
+		if _, lerr := errLogFile.WriteString(fmt.Sprintf("Error creating Kubernetes client: %v\n", err)); lerr != nil {
+			err = fmt.Errorf("%v+ error writing to stderr log file: %v", err, lerr)
+		}
+		return dc, err
+	}
+	dc.clientset = clientset
+
+	return dc, nil
+}
+
+func (d *DiagnosticCollector) Close() {
+	d.logFile.Close()
+	d.errLogFile.Close()
+}
+
+func (d *DiagnosticCollector) Add(f collectFunc) {
+	d.collectors = append(d.collectors, f)
+}
+
+func (d *DiagnosticCollector) AddExtension(extension string) {
+	switch extension {
+	case NodeFeature:
+		d.Add(d.NodeFeatures)
+	case NodeFeatureRule:
+		d.Add(d.NodeFeatureRules)
+	}
+}
+
+// ParseObjects reads from a comma separated string and sets the DiagnosticCollector objects
+// This is a helper function to set the DiagnosticCollector objects from a comma separated string
+// of objects, for making it easier to set the objects to collect when manually running the
+// diagnostic collector.
+func (d *DiagnosticCollector) ParseObjects(objects string) {
+	for _, obj := range strings.Split(objects, ",") {
```

Comment:
Is there a reason that we don't just call `d.Add()` here? What is the benefit of storing a set of strings and THEN converting that to collectors?

Reply:
Also, do we need to deduplicate the objects?

Reply:
yes, good catch, users could by mistake have something like `pods,pods,namespaces`